R/utils.R

```diff
@@ -92,3 +92,166 @@ browse_url <- function(path, chromote) {
     utils::browseURL(url)
   }
 }
+
+
+# =============================================================================
+# Random Ports
+# =============================================================================
+#
+# Borrowed from https://github.com/rstudio/httpuv/blob/main/R/random_port.R
+
+#' Startup a service that requires a random port
+#'
+#' `with_random_port()` provides `startup()` with a random port value and runs
+#' the function:
+#'
+#' 1. `startup()` always returns a value if successful.
+#' 2. If `startup()` fails with a generic error, we assume the port is occupied
+#'    and try the next random port.
+#' 3. If `startup()` fails with an error classed with errors in `stop_on`,
+#'    (`error_stop_port_search` or `system_command_error` by default), we stop
+#'    the port search and rethrow the fatal error.
+#' 4. If we try `n` random ports, the port search stops with an informative
+#'    error that includes the last port attempt error.
+#'
+#' @param startup A function that takes a `port` argument, where `port` will be
+#'   randomly selected. When successful, `startup()` should return a non-NULL
+#'   value that will also be returned from `with_random_port()`. Generic errors
+#'   emitted by this function are silently ignored: when `startup()` fails, we
+#'   assume the port was unavailable and we try with a new port. Errors with the
+#'   classes in `stop_on` fail immediately.
+#' @param ... Additional arguments passed to `startup()`.
+#' @param min,max Port range
+#' @param n Maximum number of ports to try
+#' @param stop_on Error classes that signal the port search should be stopped
+#'
+#' @return The result of `startup()`, or an error if `startup()` fails.
+#' @noRd
+with_random_port <- function(
+  startup,
+  ...,
+  min = 1024L,
+  max = 49151L,
+  n = 10,
+  stop_on = c("error_stop_port_search", "system_command_error")
+) {
+  stopifnot(is.function(startup))
+  valid_ports <- setdiff(seq.int(min, max), unsafe_ports)
+
+  # Try up to n ports
+  n <- min(n, length(valid_ports))
+  ports <- sample(valid_ports, n)
+  err_port <- NULL
+
+  .empty_result <- structure(list(), class = "no value returned from startup")
+
+  for (port in ports) {
+    res <- .empty_result
```

Comment:
Is there any particular reason not to just use `NULL` here? When I see how `.empty_result` is defined and used, it makes me look around for something really unusual in the code, and it makes me worry that I've missed something important.

 If it's because `startup()` could succeed and return `NULL`, then I think it would make sense to just define a variable like `succeeded <- FALSE`, and then set it to `TRUE` after calling `startup()`, as in:

```R
tryCatch({
    res <- startup(port = port, ...)
    succeeded <- TRUE
  },
  ...
)
```

Reply:
Yes that would be a very straight-forward approach! (And no, you didn't miss anything.) Implemented in bf9ef5f596adb1fd696712ce9c8ef958bcf08dd2.